Yet another State Machine

DeivSky opened the pull request:
Thread: https://forum.unity.com/threads/character-state-machine.979371/
Card: https://open.codecks.io/unity-open-project-1/decks/15/card/157-state-machine

Here's a post about my reasoning behind this implementation: https://forum.unity.com/threads/character-state-machine.979371/page-3#post-6422979

Basically, more than having my implementation be _the_ one we end up using, the main purpose of this **PR** is to show you how we can separate runtime components from the designer friendly components. By taking this approach, we can easily make custom **state machines** with **pluggable** components (just like any other SO-based implementations), but without the downsides that come from the shared nature of **SO**s, namely data management.

In this **PR** you'll find, besides my implementation, a super basic example of what I mean with this separation. There's the player, and 5 instances of a Sphere prefab. The Spheres start with the **Chasing** state. This state has two actions: **Chase** the player, and **EnableHunger**, which makes the spheres get hungry. There are two possible transitions from this state: either they catch the player, or their hunger gets past their set threshold. These are the **Conditions** for those two **transitions**. When met, they set the **state** of the sphere to **Attacking** or **Eating**, respectively. Both after **Attacking** and **Eating** the spheres go back to **Chasing** after a set duration (to simulate animations). 

When you enter the **Eating** state, you **eat** and recover by a set amount, and then you go from **EatingToChasing** when the specified **EatDuration** has passed.
![image](https://user-images.githubusercontent.com/27371812/96325540-0632c600-0fff-11eb-9024-0067c2188551.png)
**Note:** a custom editor is desperately needed but out of the scope of this PR.
 
These are all **ScriptableObjects** however, and, as an example, let's take the **EatDuration** condition, where would we store the amount of time that's passed for each **StateMachine**? We could store it in the **StateMachine** itself in a Dictionary<SomeKey, object> that holds the data of all the states, or in a static Dictionary<StateMachine, float> in the same **ScriptableObject** so there's no boxing, or make everything generic and bind everything to **DataModels**. But they're all workarounds for something you'd normally solve with just one simple float field inside your class. So that's what I wanted to achieve. If we had separate instances of states, actions, etc, for each state machine, then each instance could hold its own data.

And here is how it works so far:

We have **ScriptableState, ScriptableStateAction, ScriptableStateTransition and ScriptableStateCondition**. These are SOs, and each of them has a **GetState/Action/Transition/Condition** method, which returns a **State/StateAction/StateTransition/StateCondition**. The latter being plain C# scripts, and they contain the actual implementation. 

The **StateMachine**, on Awake, initiates a chain reaction when calling **GetState** on the initial State SO. It's a chain reaction because, when getting the state, we're also calling **GetAction** and **GetTransition** for each of the actions and transitions in it. The transitions would then call **GetCondition** for all of their conditions, and **GetState** for their target state, repeating the process again and again, until the whole behaviour tree made with SOs is resolved into its runtime counterparts. We use a Dictionary<ScriptableObject, object> during this whole process to keep track of the already created instances in order to not create again.

And that's the main idea. Implementation details such as using **TransitionTables** or having a **NodeGraph** are not discarded. Rather, I think taking this idea and apply it to any SO workflow we end up using would be beneficial.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/UnityTechnologies/open-project-1?pullRequest=94) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/UnityTechnologies/open-project-1?pullRequest=94) before we can accept your contribution.<br/><hr/>**DeivSky** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/UnityTechnologies/open-project-1?pullRequest=94) it.</sub>

ciro-unity commented on UOP1_Project/Assets/Scripts/StateMachineTest/ScriptableHungerAction.cs:
```diff
@@ -0,0 +1,19 @@
+{
+	private HungerComponent _hungerComponent;
+
+	public override void Awake(StateMachine stateMachine)
+		=> _hungerComponent = stateMachine.GetComponent<HungerComponent>();
+
+	public override void Perform() { }
+
+	public override void OnStateEnter() => _hungerComponent.ToggleHunger(true);
+	public override void OnStateExit() => _hungerComponent.ToggleHunger(false);
```
Why did you make all functions into lambdas?

DeivSky replied:
Not lambdas, expression body definitions. 
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/lambda-operator

It's exactly the same as writing
``` csharp
public override void OnStateEnter()
{
    _hungerComponent.ToggleHunger(true);
}
```
Just less boilerplate.

ciro-unity replied:
Yeah, I don't know, I think it's quite confusing to have half of the codebase with methods expressed with curly braces, and half of it using expression body definitions.
Personally I would avoid them, I generally like the "framing" that curly braces give to a function's body (even if it's a one liner), and they are also kind of obscure for beginner users.

DeivSky replied:
Yeah it's personal preference. I did ask about it in the coding standards thread a while ago, no one said much about it though. I don't mind switching to brackets, should we add that to the document?

adamhill replied:
Expression bodied members have been around for half a decade (C# 6 - 2015) hardly obscure. One of the most maddening things about Unity examples is how they seem to refuse to use new, simple syntax idioms that reduce visual noise and make things easier to read. Pattern matching is the next hill that Unity evangelists are going to have to climb and need to push into common usage.

Part of this used to come from Unity being so far behind (pre 3.5) Mono releases, so old syntax was commonplace (like "get {}; set{}; instead of "{get;set}"  for simple Props) but now that Unity is going to track C# at a quicker pace, Unity "best practices" should keep up with modern usage if possible.

Using R# or Rider it is simple to switch everything to expressions bodied methods for consistency.

ciro-unity replied:
> simple syntax idioms that reduce visual noise and make things easier to read

I don't know if I find expression-bodied members easier to read, less characters don't necessarily mean more readable. In fact I find them less readable!

But as @DeivSky said, it's a matter of preference so I don't mind too much if he wants to keep them in the state machine, but I personally wouldn't love to see them all over the codebase.

ciro-unity commented on UOP1_Project/Assets/Scripts/StateMachineTest/ScriptableCloseToTargetCondition.cs:
```diff
@@ -0,0 +1,39 @@
+using UnityEngine;
+
+[CreateAssetMenu(fileName = "CloseToTarget", menuName = "State Machines/Tests/Conditions/Close To Target")]
+public class ScriptableCloseToTargetCondition : ScriptableStateCondition
```
All throughout the implementation you're using "Scriptable" as a prefix for classes that represent ScriptableObjects. 

But "scriptable" just means that they can contain some behaviour, in this sense MonoBehaviours are scriptable too, and it makes it all very confusing.
I believe it would be better to use the prefix "ScriptableObject", or "SO", to indicate that they are Scriptable Objects.

For instance, this class would be `SOCloseToTargetCondition` or (my preference) `CloseToTargetConditionSO`.

DeivSky replied:
I agree and I admit I'm not particularly good with naming. I like SO at the end, I'll start using it!

ciro-unity replied:
You don't need to bash yourself :D

ciro-unity commented on UOP1_Project/Assets/Scripts/StateMachineTest/ChaseDataObject.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using UnityEngine;
+
+[CreateAssetMenu(fileName = "ChaseData", menuName = "State Machines/Tests/Data/Chase Data")]
+public class ChaseDataObject : ScriptableObject
```
Similar consideration to the other comment about the "Scriptable" prefix. Here `ChaseDataObject` should really be `ChaseDataSO` for clarity on what type of object it is?

ciro-unity reviewed (commented):
Left a few comments where I wasn't sure about functionality or have some suggestions to improve understandability.

DeivSky commented on UOP1_Project/Assets/Scripts/StateMachine/Core/StateMachine.cs:
```diff
@@ -0,0 +1,106 @@
+#endif
+		}
+
+		private static List<Type> GetObjectTypes(object[] objects)
```
I had it with SO[] at first but then I thought, this function works for any type of object so it doesn't really have to be restricted to SOs.

ciro-unity replied:
But does it need to? (work with any type of object)

DeivSky replied:
I guess not 👍🏻

DeivSky commented on UOP1_Project/Assets/Scripts/StateMachine/Core/StateMachine.cs:
```diff
@@ -0,0 +1,106 @@
+		public bool debug;
+#endif
+		[SerializeField] private Scriptables.ScriptableState _initialStateSO = null;
+		[SerializeField] private ScriptableObject[] _scriptableObjects = null;
```
Yeah that's sort of what it is. The idea was to be able to override parameters of an Action/Condition.
It's used in the `ChaseAction` in my example. The asset takes in a default `ChaseDataObject` (or `ChaseDataSO`, I'll rename everything to that format)
``` csharp
[CreateAssetMenu(fileName = "Chase", menuName = "State Machines/Tests/Actions/Chase")]
public class ScriptableChaseAction : ScriptableStateAction
{
	[SerializeField] private ChaseDataObject _chaseData = null;

	protected override StateAction CreateAction() => new ChaseAction(_chaseData);
}
```

That would be the default `ChaseDataSO`. However, on `Awake`, the action checks if there's a different one in the `StateMachine`.
``` csharp
```

ciro-unity replied:
Isn't this a bit like overthinking? To recap: There's already the ability to assign a different action class, and for each "same" action we can customise its parameters by assigning a different SO to it containing different data, right? (correct me if I'm wrong)

Is it needed to also be able to slightly modify the behaviour of a certain action based on which StateMachine is using it...? Isn't it better to just create a different action altogether? I think this feels a bit like overthinking, introducing a lot of complexity and extra methods all over the FSM.

ciro-unity replied:
Ah, I see it a bit better now. So basically that array allows state machines to have their own unique data per instance, right?

I guess it's useful then. Ignore my previous comments. But I would still name it to suggest that... `_overrideData` or `_uniqueData`?

ciro-unity replied:
The annoying thing with this implementation is that this data provided via the array is not tied to the state/action/condition it belongs to. So for instance, having 2 SOs of the same type in this array means you don't know which one is used in the end.
On the other hand if you have 2 state/action/condition that require the same type of data, you have no option to provide 2 different SOs to these... they will both pick up the same one from the array.

A great way would be that whenever you add a state/action/condition to the state machine, this surfaces the field to feed the data (SO) it needs to work to the Inspector of the StateMachine component. In this respect, the Inspector would end up looking a lot like the one Neonage created in #51, but without the logic.

ciro-unity replied:
One final comment: I feel like you were somehow aware of this issue while coding the whole thing and in fact you broke the pattern you set with the `HungerComponent.cs`, which acts as a container of data unique to the instance of the StateMachine, in fact it's the only non-ScriptableObject script in the StateMachineTest folder.

I guess in a way it shows the flexibility of your implementation (or of Unity? :D) but at the same time it can be confusing: when should I use the array of overrides (`_scriptableObjects`) and when instead use a MonoBehaviour and just dump the data in there?

DeivSky replied:
>The annoying thing with this implementation is that this data provided via the array is not tied to the state/action/condition it belongs to. So for instance, having 2 SOs of the same type in this array means you don't know which one is used in the end.
On the other hand if you have 2 state/action/condition that require the same type of data, you have no option to provide 2 different SOs to these... they will both pick up the same one from the array.

I guess I thought, if multiple action/condition use the same type of data, it's likely because they share it, and my idea was to somehow restrict adding two SOs of the same type. Now that you say it though, that's probably not always the case. Unless we make different types for everything, but that wouldn't make much sense.

>A great way would be that whenever you add a state/action/condition to the state machine, this surfaces the field to feed the data (SO) it needs to work to the Inspector of the StateMachine component. In this respect, the Inspector would end up looking a lot like the one Neonage created in #51, but without the logic.

Definitely. I'm really bad at Editor scripting and as I said it was out of scope, but that would be a great way to handle it.

>One final comment: I feel like you were somehow aware of this issue while coding the whole thing and in fact you broke the pattern you set with the HungerComponent.cs, which acts as a container of data unique to the instance of the StateMachine, in fact it's the only non-ScriptableObject script in the StateMachineTest folder.

Yup, you got me! I had to access `HungerComponent` from `IsHungryCondition` and `EnableHungerAction`, also it had to be per instance data, so using SOs made no sense. I was going to overcomplicate it even further by making `HungerComponent` a regular C# class and having another dictionary in the state machine, `Dictionary<Type, object>`, and a `public T GetOrAddObject<T>() where T : class, new()`, this way we could have objects shared between SOs... 
I know, you weren't wrong, I overthink too much :'D. I scraped that and opted for a MonoBehaviour. 

>I guess in a way it shows the flexibility of your implementation (or of Unity? :D) 

I thought about that too while doing that!

>but at the same time it can be confusing: when should I use the array of overrides (_scriptableObjects) and when instead use a MonoBehaviour and just dump the data in there?

I'm trying to get an answer but I seem to be leaning towards removing the array of overrides and just using MonoBehaviours. The array adds an unnecessary layer of complexity, and everything done with it could easily be made with MBs. I think my main problem was that I was thinking of SOs as replacers of MBs, that's why I had exposed values in some actions/conditions. But SOs should simply call the methods in MBs, leaving it up to them however they get configured.

DeivSky replied:
This has been really insightful, thank you for all your input!

ciro-unity replied:
No problem. Do you have an idea of how to improve it then? I would like to merge the changes if we get to a good solution.

DeivSky replied:
In terms of the implementation itself, I think just removing the 'override' feature would do. Everything else is specific to the test scene. 
Then a custom editor is definitely needed, and I'd also like to change how transitions are made, maybe have transition tables like kcastagnini in his [#76](https://github.com/UnityTechnologies/open-project-1/pull/76), but a lot of that comes from editor scripting as well and, as I said, I'm not that good at it, so I'd need some help with that.

DeivSky replied:
Alright, I removed the `_overrideData` and I'm liking it better now, it's a lot cleaner. I think it's in a decent state now, do you see any other improvements -not related to the inspector- I could work on?

DeivSky commented on UOP1_Project/Assets/Scripts/StateMachineTest/ScriptableTimedCondition.cs:
```diff
@@ -0,0 +1,23 @@
+	protected override Condition CreateCondition() => new TimedCondition(Duration);
+}
+
+public class TimedCondition : Condition
```
Yeah I was hesitant about this one. I certainly don't like how the 'd' looks there for some reason either, I just thought Timer didn't really fit with it being a Condition. I really need to work on my naming.

DeivSky commented on UOP1_Project/Assets/Scripts/StateMachine/Core/StateAction.cs:
```diff
@@ -0,0 +1,22 @@
+	/// </summary>
+	public abstract class StateAction : IStateComponent
+	{
+		/// <summary>
+		/// Perform the action this object represents.
+		/// </summary>
+		public abstract void Perform();
```
I used that name because at first I didn't have `OnStateEnter` and `OnStateExit` but rather three sets of `StateAction[]` in the `State`, them being `_onStateEnterActions`, `_onStateExitActions`, and `_onUpdateActions` that the `State` would call in their respective times. So I needed a generic name for it. When I switched formats I just kept the name. I'll change it to `OnUpdate`.

DeivSky commented on UOP1_Project/Assets/Scripts/StateMachine/Core/StateMachine.cs:
```diff
@@ -0,0 +1,106 @@
+		}
+	}
+
+	public abstract class StateMachine<T> : StateMachine
+	{
+		public abstract T GetContext();
+	}
```
At some point I was going to implement sort of what **kcastagnini** did with his version, making everything generic so that every `Action<T>`, `Condition<T>`, `State<T>`, etc, can only live in `StateMachine<T>` of the same `<T>`. I ended up not doing it and I should've removed that.

ciro-unity commented:
Merging this, although I need to do a bit extra work to integrate some changes which I made locally, and now is not allowing me to push to your branch @DeivSky 🤷🏻‍♂️

ciro-unity commented:
Managed to merge, check out `UnityTechnologies:main` for the very latest changes.

A few things:
- I renamed the package from Devisky.StateMachine to UOP1.StateMachine, we want to keep generic naming so I'll put that in the guidelines too
- I removed the example test, now that we have the character state machine implemented I felt like it was not needed. Same for the sample scene.
- I ended up "solving" the SO vs. MonoBehaviour issue by using a MB as a data container on the instance. It's called `Character.cs`. Check it out in the code by pulling the newest one from `main`.
- Something else I don't remember!

ciro-unity commented:
Ah yes, are you on Discord? If so, tell me your complete nickname and I can give you the role of "Open Project Contributor". Same for @adamhill
This is different from the credits in the game, where we'll be using your real name taken from Github or the username if not possible.
No worries if you don't care about Discord.

DeivSky commented:
> Managed to merge, check out `UnityTechnologies:main` for the very latest changes.
> 
> A few things:
> 
> * I renamed the package from Devisky.StateMachine to UOP1.StateMachine, we want to keep generic naming so I'll put that in the guidelines too
> * I removed the example test, now that we have the character state machine implemented I felt like it was not needed. Same for the sample scene.
> * I ended up "solving" the SO vs. MonoBehaviour issue by using a MB as a data container on the instance. It's called `Character.cs`. Check it out in the code by pulling the newest one from `main`.
> * Something else I don't remember!

Awesome, I'm checking it out right now! Although I think there was some issue, everything from the test scene is still there :S

DeivSky commented:
> Ah yes, are you on Discord? If so, tell me your complete nickname and I can give you the role of "Open Project Contributor". Same for @adamhill
> This is different from the credits in the game, where we'll be using your real name taken from Github or the username if not possible.
> No worries if you don't care about Discord.

I am! It's DavidL#1006
